types/wlr_keyboard.c

```diff
@@ -13,19 +13,45 @@ int os_create_anonymous_file(off_t size);
 static void keyboard_led_update(struct wlr_keyboard *keyboard) {
 	uint32_t leds = 0;
 	for (uint32_t i = 0; i < WLR_LED_LAST; ++i) {
-		if (xkb_state_led_index_is_active(keyboard->xkb_state, keyboard->leds[i])) {
+		if (xkb_state_led_index_is_active(keyboard->xkb_state,
+				keyboard->led_indexes[i])) {
 			leds |= (1 << i);
 		}
 	}
 	wlr_keyboard_led_update(keyboard, leds);
 }
 
+static void keyboard_modifier_update(struct wlr_keyboard *keyboard) {
+	xkb_mod_mask_t depressed = xkb_state_serialize_mods(keyboard->xkb_state,
+		XKB_STATE_DEPRESSED);
+	xkb_mod_mask_t latched = xkb_state_serialize_mods(keyboard->xkb_state,
+		XKB_STATE_LATCHED);
+	xkb_mod_mask_t locked = xkb_state_serialize_mods(keyboard->xkb_state,
+		XKB_STATE_LOCKED);
+	xkb_mod_mask_t group = xkb_state_serialize_layout(keyboard->xkb_state,
+		XKB_STATE_LAYOUT_EFFECTIVE);
+	if (depressed == keyboard->modifiers.depressed &&
+			latched == keyboard->modifiers.latched &&
+			locked == keyboard->modifiers.locked &&
+			group == keyboard->modifiers.group) {
+		return;
+	}
+
+	keyboard->modifiers.depressed = depressed;
+	keyboard->modifiers.latched = latched;
+	keyboard->modifiers.locked = locked;
+	keyboard->modifiers.group = group;
+
+	// TODO: wl_keyboard_send_modifiers
```

Comment:
This isn't your job, it's already being handled by the seat. But maybe you should look for other places modifiers are being handled and attempt to deduplicate them?

Reply:
Oh, right, thanks! (wlr_seat.c:405)